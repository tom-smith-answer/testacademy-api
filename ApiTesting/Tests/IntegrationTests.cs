namespace ApiTesting.Tests
{
    using System;
    using static RestAssured.Dsl;
    using FluentAssertions;
    using Microsoft.AspNetCore.Builder;
    using Microsoft.Extensions.Hosting;
    using Microsoft.AspNetCore.Mvc.Testing;
    using Microsoft.AspNetCore.Hosting;
    using Dotnet.AspNetCore.Samples.WebApi;
    using System.Security.AccessControl;
    using System.Runtime.CompilerServices;
    using Newtonsoft.Json.Linq;
    using Snapshooter.NUnit;
    using RestAssured.Response;
    using ApiTesting.Classes.TestData;
    using FluentAssertions.Equivalency;
    using ApiTesting.Classes.POCO;
    using ApiTesting.Classes.Utilities;

    public class IntegrationTests : BasicTests
    {
        [TearDown]
        public void TearDown()
        {
            Given(httpClient)
                .When()
                .Delete(baseUrl + "Players/13")
                .Then();
        }

        [TestCase("Players", 200)]
        [TestCase("Players/1", 200)]
        [TestCase("Players/999", 404)]
        [TestCase("Players/fail", 400)]
        public void GetStatusCode(string endpoint, int status)
        {
            Given(httpClient)
                .When()
                .Get(baseUrl + endpoint)
                .Then()
                .StatusCode(status);
        }

        [TestCase("Players")]
        [TestCase("Players/1")]
        public void GetContentType(string endpoint)
        {
            Given(httpClient)
            .When()
            .Get(baseUrl + endpoint)
            .Then()
            .StatusCode(200)
            .ContentType("application/json; charset=utf-8");
        }

        [TestCase("Players")]
        public void GetPlayersBodyContent(string endpoint)
        {
            object playerResponse = Given(httpClient)
                .When()
                .Get(baseUrl + endpoint)
                .Then()
                .StatusCode(200)
            .Extract().Body("$");

            Snapshot.Match(playerResponse);
        }

        [Test]
        public void GetPlayerBodyContent()
        {
            object playerResponse = Given(httpClient)
                .When()
                .Get(baseUrl + "Players/1")
                .Then()
                .StatusCode(200)
            .Extract().Body("$");

            Snapshot.Match(playerResponse);
        }

        [Test]
        public void GetResponseTime()
        {
            Given(httpClient)
            .When()
            .Get(baseUrl + "Players")
            .Then()
            .ResponseTime(NHamcrest.Is.LessThan(TimeSpan.FromMilliseconds(2000)));
        }

        [TestCase("Post_Harvey", 201, "id")]
        [TestCase("Post_Incomplete", 400, "traceId")]
        public void PostRequest(string name, int status, string ignore)
        {
            DataHelper dataHelper = new DataHelper();

            Player player = dataHelper.GetPlayerData(name);

         object playerResponse = Given(httpClient)
        .Body(player)
        .When()
        .Post(baseUrl + "Players")
        .Then()
        .StatusCode(status)
        .Extract().Body("$");

         Snapshot.Match(playerResponse, matchOptions => matchOptions.IgnoreField(ignore));
        }

        [Test]
        public void PostEmptyRequest()
        {
            Given(httpClient)
                .Body("")
                .When()
                .Post(baseUrl + "Players")
                .Then().StatusCode(400);
        }

        [Test]
        public void PostInvalidRequest () 
        {
            InvalidPlayer player = new InvalidPlayer 
            {
                Id = 0,
                FirstName = 1,
                MiddleName = 2,
                LastName = 3,
                DateOfBirth = 4,
                SquadNumber = "5",
                Position = 6, 
                AbbrPosition = 7,
                Team = 8,
                League = 9,
                Starting11 = "true"
            };

            object playerResponse = Given(httpClient)
                .Body(player)
                .When()
                .Post(baseUrl + "Players")
                .Then()
                .StatusCode(400)
                .Extract().Body("$");

            Snapshot.Match(playerResponse, matchOptions => matchOptions.IgnoreField("traceId"));
        }

        [TestCase("Players/13", 204)]
        [TestCase("Players/14", 404)]
        [TestCase("Players/fail", 400)]
        public void DeletePlayerCheckStatus(string endpoint, int status)
        {
            DataHelper dataHelper = new DataHelper();

            Player player = dataHelper.GetPlayerData("Post_Duncan");

            Given(httpClient)
                .Body(player)
                .When()
                .Post(baseUrl + "Players")
                .Then();

            Given(httpClient)
                .When()
                .Delete(baseUrl + endpoint)
                .Then()
                .StatusCode(status);
        }

        [TestCase("Update_Harvey_Invalid", 400)]
        [TestCase("Update_Harvey_Incomplete", 400)]
        [TestCase("Update_Harvey", 204)]
        [TestCase("Update_Harvey_Multi", 204)]
        public void PutPlayer(string name, int status)
        {
            DataHelper dataHelper = new DataHelper();
            Player player = dataHelper.GetPlayerData(name);

            Given(httpClient)
                .Body(player)
                .When()
                .Put(baseUrl + "Players/12")
                .Then()
                .StatusCode(status);

            object playerResponse = Given(httpClient)
                .When()
                .Get(baseUrl + "Players/12")
                .Then()
                .Extract().Body("$");


            Snapshot.Match(playerResponse);
        }

        [TestCase("Players/13", 400)]
        [TestCase("Players/fail", 400)]

        public void PutBadEndpoint(string endpoint, int status)
        {
            DataHelper dataHelper = new DataHelper();
            Player player = dataHelper.GetPlayerData("Update_Harvey");

            Given(httpClient)
                .Body(player)
                .When()
                .Put(baseUrl + endpoint)
                .Then()
                .StatusCode(status);
        }

        [Test]
        public void PutEmpty()
        {
            Given(httpClient)
                .Body("")
                .When()
                .Put(baseUrl + "Players/12")
                .Then()
                .StatusCode(400);
        }
    }
}

