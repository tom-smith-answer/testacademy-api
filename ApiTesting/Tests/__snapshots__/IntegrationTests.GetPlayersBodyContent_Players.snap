[
  {
    "id": 1,
    "firstName": "Damián",
    "middleName": "Emiliano",
    "lastName": "Martínez",
    "dateOfBirth": "1992-09-02T01:00:00+01:00",
    "squadNumber": 23,
    "position": "Goalkeeper",
    "abbrPosition": "GK",
    "team": "Aston Villa FC",
    "league": "Premier League",
    "starting11": false
  },
  {
    "id": 2,
    "firstName": "Nahuel",
    "middleName": null,
    "lastName": "Molina",
    "dateOfBirth": "1998-04-06T01:00:00+01:00",
    "squadNumber": 26,
    "position": "Right-Back",
    "abbrPosition": "RB",
    "team": "Atlético Madrid",
    "league": "La Liga",
    "starting11": false
  },
  {
    "id": 3,
    "firstName": "Cristian",
    "middleName": "Gabriel",
    "lastName": "Romero",
    "dateOfBirth": "1998-04-27T01:00:00+01:00",
    "squadNumber": 13,
    "position": "Centre-Back",
    "abbrPosition": "CB",
    "team": "Tottenham Hotspur",
    "league": "Premier League",
    "starting11": false
  },
  {
    "id": 4,
    "firstName": "Nicolás",
    "middleName": "Hernán Gonzalo",
    "lastName": "Otamendi",
    "dateOfBirth": "1988-02-12T00:00:00+00:00",
    "squadNumber": 19,
    "position": "Centre-Back",
    "abbrPosition": "CB",
    "team": "SL Benfica",
    "league": "Liga Portugal",
    "starting11": false
  },
  {
    "id": 5,
    "firstName": "Nicolás",
    "middleName": "Alejandro",
    "lastName": "Tagliafico",
    "dateOfBirth": "1992-08-31T01:00:00+01:00",
    "squadNumber": 3,
    "position": "Left-Back",
    "abbrPosition": "LB",
    "team": "Olympique Lyon",
    "league": "Ligue 1",
    "starting11": false
  },
  {
    "id": 6,
    "firstName": "Ángel",
    "middleName": "Fabián",
    "lastName": "Di María",
    "dateOfBirth": "1988-02-14T00:00:00+00:00",
    "squadNumber": 11,
    "position": "Right Winger",
    "abbrPosition": "LW",
    "team": "SL Benfica",
    "league": "Liga Portugal",
    "starting11": false
  },
  {
    "id": 7,
    "firstName": "Rodrigo",
    "middleName": "Javier",
    "lastName": "de Paul",
    "dateOfBirth": "1994-05-24T01:00:00+01:00",
    "squadNumber": 7,
    "position": "Central Midfield",
    "abbrPosition": "CM",
    "team": "Atlético Madrid",
    "league": "La Liga",
    "starting11": false
  },
  {
    "id": 8,
    "firstName": "Enzo",
    "middleName": "Jeremías",
    "lastName": "Fernández",
    "dateOfBirth": "2001-01-17T00:00:00+00:00",
    "squadNumber": 24,
    "position": "Central Midfield",
    "abbrPosition": "CM",
    "team": "Chelsea FC",
    "league": "Premier League",
    "starting11": false
  },
  {
    "id": 9,
    "firstName": "Alexis",
    "middleName": null,
    "lastName": "Mac Allister",
    "dateOfBirth": "1998-12-24T00:00:00+00:00",
    "squadNumber": 20,
    "position": "Central Midfield",
    "abbrPosition": "CM",
    "team": "Liverpool FC",
    "league": "Premier League",
    "starting11": false
  },
  {
    "id": 10,
    "firstName": "Lionel",
    "middleName": "Andrés",
    "lastName": "Messi",
    "dateOfBirth": "1987-06-24T01:00:00+01:00",
    "squadNumber": 10,
    "position": "Right Winger",
    "abbrPosition": "RW",
    "team": "Inter Miami CF",
    "league": "Major League Soccer",
    "starting11": false
  },
  {
    "id": 11,
    "firstName": "Julián",
    "middleName": null,
    "lastName": "Álvarez",
    "dateOfBirth": "2000-01-31T00:00:00+00:00",
    "squadNumber": 9,
    "position": "Centre-Forward",
    "abbrPosition": "CF",
    "team": "Manchester City",
    "league": "Premier League",
    "starting11": false
  },
  {
    "id": 12,
    "firstName": "Harvey",
    "middleName": "The Power",
    "lastName": "Sembhy",
    "dateOfBirth": "1066-06-06T14:30:23.234",
    "squadNumber": 69,
    "position": "Everywhere",
    "abbrPosition": "EW",
    "team": "Ninja Warriors",
    "league": "Answer under 1100s",
    "starting11": false
  }
]
